src/main/php/xp/web/srv/ForwardRequests.class.php

```diff
@@ -0,0 +1,103 @@
+<?php namespace xp\web\srv;
+
+use peer\Socket;
+use web\io\{ReadChunks, ReadLength};
+
+/**
+ * Forwards HTTP requests to the given backend
+ *
+ * @test  web.unittest.server.ForwardRequestsTest
+ */
+class ForwardRequests extends Switchable {
+  private $backend;
+
+  /** Creates a new instance */
+  public function __construct(Socket $backend) {
+    $this->backend= $backend;
+  }
+
+  /**
+   * Transmits data from an optional stream to a given target socket,
+   * handling chunked transfer-encoding.
+   *
+   * @see    https://developer.mozilla.org/de/docs/Web/HTTP/Headers/Transfer-Encoding
+   * @param  io.streams.InputStream $stream
+   * @param  peer.Socket $target
+   * @return void
+   */
+  private function transmit($stream, $target) {
+    if (null === $stream) {
+      // NOOP
+    } else if ($stream instanceof ReadChunks) {
+      while ($stream->available()) {
+        yield;
+        $chunk= $stream->read();
+        $target->write(dechex(strlen($chunk))."\r\n".$chunk."\r\n");
+      }
+      $target->write("0\r\n\r\n");
+    } else {
+      while ($stream->available()) {
+        yield;
+        $target->write($stream->read());
+      }
+    }
+  }
+
+  /**
+   * Handle client data
+   *
+   * @param  peer.Socket $socket
+   * @return void
+   */
+  public function handleData($socket) {
+    static $exclude= ['Remote-Addr' => true];
+
+    $request= new Input($socket);
+    yield from $request->consume();
+
+    if (Input::REQUEST === $request->kind) {
+      $this->backend->connect();
+      $message= "{$request->method()} {$request->uri()} HTTP/{$request->version()}\r\n";
+      $headers= [];
+      foreach ($request->headers() as $name => $value) {
+        isset($exclude[$name]) || $message.= "{$name}: {$value}\r\n";
+        $headers[$name]= $value;
+      }
+      // \util\cmd\Console::writeLine('>>> ', $message);
+      $this->backend->write($message."\r\n");
+      foreach ($this->transmit($request->incoming(), $this->backend) as $step) {
+        // yield 'read' => $socket;
```

Comment:
This is commented to guarantee synchronous execution of this method due to the backend server only being able to handle one request at a time. We could consider using a worker pool in order to speed things up in a future pull request, or handling requests to static resources inside the proxy itself.

Reply:
> We could consider using a worker pool

Initial setup done in xp-forge/web@c5414d9